File: 1171.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

#include <unordered_map>

struct Solution {
    ListNode* removeZeroSumSublists(ListNode* head) {
        bool removed = true;
        unordered_map<int, ListNode*> d;
        while (removed){
            removed = false;
            ListNode foobar(0, head);
            d.clear();
            d.insert({0, &foobar});
            int s = 0;
            ListNode *ptr = head;
            while (ptr != nullptr){
                s += ptr->val;
                if (d.contains(s)){
                    d[s]->next = ptr->next;
                    removed = true;
                }
                d[s] = ptr;
                ptr = ptr->next;
            }
            head = foobar.next;
        }
        return head;
    }
};

File: sol/1171/1171.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        removed = True
        while removed:
            removed = False
            foobar = ListNode(0, head)
            d = {0: foobar}
            ptr = head
            s = 0
            while ptr is not None:
                s += ptr.val
                if s in d:
                    d[s].next = ptr.next
                    removed = True
                d[s] = ptr
                ptr = ptr.next
            head = foobar.next
        return head
